x-pack/auditbeat/module/system/login/login.go

```diff
@@ -0,0 +1,234 @@
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+	}
+
+	ms.utmpReader, err = NewUtmpFileReader(ms.log, bucket, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	return ms.utmpReader.Close()
+}
+
+// Fetch collects any new login records from /var/log/wtmp. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	count := ms.readAndEmit(report)
+
+	ms.log.Debugf("%d new login records.", count)
+
+	// Save new state to disk
+	if count > 0 {
+		err := ms.utmpReader.saveStateToDisk()
+		if err != nil {
+			ms.log.Error(err)
+			report.Error(err)
+		}
+	}
+}
+
+// readAndEmit reads and emits login events and returns the number of events.
+func (ms *MetricSet) readAndEmit(report mb.ReporterV2) int {
+	loginRecordC, errorC := ms.utmpReader.ReadNew()
+
+	var count int
+	for {
+		select {
+		case loginRecord, ok := <-loginRecordC:
+			if !ok {
+				return count
+			}
+			report.Event(ms.loginEvent(&loginRecord))
+			count++
+		case err, ok := <-errorC:
+			if !ok {
+				return count
+			}
+			ms.log.Error(err)
+		}
+	}
+}
+
+func (ms *MetricSet) loginEvent(loginRecord *LoginRecord) mb.Event {
+	event := mb.Event{
+		Timestamp: loginRecord.Timestamp,
+		RootFields: common.MapStr{
+			"event": common.MapStr{
+				"type":   eventTypeEvent,
+				"action": loginRecord.Type.string(),
+				"origin": loginRecord.Origin,
+			},
+			"message": loginMessage(loginRecord),
+			// Very useful for development
+			// "debug": fmt.Sprintf("%v", login.Utmp),
+		},
+	}
+
+	if loginRecord.Username != "" {
+		event.RootFields.Put("user.name", loginRecord.Username)
+
+		if loginRecord.UID != -1 {
+			event.RootFields.Put("user.id", loginRecord.UID)
+		}
+	}
+
+	if loginRecord.TTY != "" {
+		event.RootFields.Put("user.terminal", loginRecord.TTY)
+	}
+
+	if loginRecord.PID != -1 {
+		event.RootFields.Put("process.pid", loginRecord.PID)
+	}
+
+	if loginRecord.IP != nil {
+		event.RootFields.Put("source.ip", loginRecord.IP)
+	}
+
+	if loginRecord.Hostname != "" && loginRecord.Hostname != loginRecord.IP.String() {
+		event.RootFields.Put("source.domain", loginRecord.Hostname)
+	}
+
+	switch loginRecord.Type {
+	case userLoginRecord:
+		event.RootFields.Put("event.outcome", "success")
+	case userLoginFailedRecord:
+		event.RootFields.Put("event.outcome", "failure")
+	}
+
+	return event
+}
+
+func loginMessage(loginRecord *LoginRecord) string {
+	var message string
+
+	userString := fmt.Sprintf("by user %v (UID: %d) on %v (PID: %d) from %v (IP: %v)",
```

Comment:
From an allocation perspective it might be better to use a `strings.Builder` + `fmt.Fprintf` or refactor so that there's only one `fmt.Sprintf`. You'll probably have to swap the order around a bit. For `bootRecord` and `shutdownRecord` you can do a straight `return "System ..."`.